cmd/telegraf/telegraf.go

```diff
@@ -231,30 +232,61 @@ func (t *Telegraf) watchLocalConfig(ctx context.Context, signals chan os.Signal,
 		return
 	}
 	log.Printf("I! Config watcher started for %s\n", fConfig)
-	select {
-	case <-ctx.Done():
-		mytomb.Done()
-		return
-	case <-changes.Modified:
-		log.Printf("I! Config file/directory %q modified\n", fConfig)
-	case <-changes.Deleted:
-		// deleted can mean moved. wait a bit a check existence
-		<-time.After(time.Second)
-		if _, err := os.Stat(fConfig); err == nil {
-			log.Printf("I! Config file/directory %q overwritten\n", fConfig)
-		} else {
-			log.Printf("W! Config file/directory %q deleted\n", fConfig)
+
+	// Setup debounce timer
+	reloadTimer := time.NewTimer(t.watchDebounceInterval)
+	reloadTimer.Stop() // stop immediately to avoid triggering reload
+	reloadPending := false
+
+	// Helper to reset the timer and mark reload as pending
+	resetTimer := func(reason string) {
+		if !reloadPending {
+			reloadPending = true
+		}
+		reloadTimer.Reset(t.watchDebounceInterval)
+		log.Printf("%s", reason)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			reloadTimer.Stop()
+			mytomb.Done()
+			return
+
+		case <-changes.Modified:
+			resetTimer(fmt.Sprintf("I! Config file/directory %q modified\n", fConfig))
+
+		case <-changes.Deleted:
+			// deleted can mean moved. wait a bit a check existence
+			<-time.After(time.Second)
+			var reason string
+			if _, err := os.Stat(fConfig); err == nil {
+				reason = fmt.Sprintf("I! Config file/directory %q overwritten\n", fConfig)
+			} else {
+				reason = fmt.Sprintf("W! Config file/directory %q deleted\n", fConfig)
+			}
+			resetTimer(reason)
+
+		case <-changes.Truncated:
+			resetTimer(fmt.Sprintf("I! Config file/directory %q truncated\n", fConfig))
+
+		case <-changes.Created:
+			resetTimer(fmt.Sprintf("I! Config directory %q has new file(s)\n", fConfig))
+
+		case <-reloadTimer.C:
+			if reloadPending {
+				log.Printf("I! Debounce period elapsed, triggering config reload for %q\n", fConfig)
+				signals <- syscall.SIGHUP
+				reloadPending = false
+			}
```

Comment:
This case executes even when debouncing is disabled when watchDebounceInterval is 0, this case should not be reachable
also, Signal send could block if channel is full `signals <- syscall.SIGHUP`
May I suggest:
```suggestion
                case <-func() <-chan time.Time {
                    if reloadTimer != nil {
                        return reloadTimer.C
                    }
                    // Return a channel that never fires when debouncing is disabled
                    return make(<-chan time.Time)
                }():
                    if reloadPending {
                        log.Printf("I! Debounce period elapsed, triggering config reload for %q\n", fConfig)
                        select {
                        case signals <- syscall.SIGHUP:
                        case <-ctx.Done():
                            return
                        }
                        reloadPending = false
                    }
```

Reply:
Ack